### PR TITLE
[DLOG] Delete empty inprogress segment on recovery

### PR DESCRIPTION
**Pull request by ivankelly**
This addresses an issue, whereby if a user is using getLastTxId() to
find the first txnid when starting to write, they will continue to be
able to write if the last segment written by the previous writer was
empty. If the last segment was empty, then the maxTxId would be higher
than the result of getLastTxId(). maxTxId is read from a znode, while
getLastTxId() reads the txid of the last persisted record. In the case
of an empty inprogress segment, the maxTxId znode was being updated
with the expected first transaction id of the segment.

This patch addresses the issue with the following changes:
1. The maxTxId znode is only updated when inprogress segment is
   completed, so it's value always refers to a transaction that
   exists.
2. On recovery, if the inprogress segment is empty, delete it. There
   was a TODO comment to do this already there.
3. When generating the sequence number, allow a potential sequence
   number which is equal to the current max sequence number, as this
   can be the case where recovery deleted an inprogress empty segment.

### CODE REVIEW

File: stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestDistributedLogBase.java [hunk 1]
```diff
@@ -95,15 +95,16 @@
...
 
     @BeforeClass
     public static void setupCluster() throws Exception {
         File zkTmpDir = IOUtils.createTempDir("zookeeper", "distrlog");
-        tmpDirs.add(zkTmpDir);
+        TMP_DIRS.add(zkTmpDir);
         Pair<ZooKeeperServerShim, Integer> serverAndPort = LocalDLMEmulator.runZookeeperOnAnyPort(zkTmpDir);
         zks = serverAndPort.getLeft();
         zkPort = serverAndPort.getRight();
+        LOG.info("IKDEBUG num bookies is {}", numBookies);
```
Comment (eolivelli) [inline stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestDistributedLogBase.java]: Nit: remove Ivan Kelly's debug?

File: stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test Cases for RollLogSegments.
+ */
+public class TestTxnId extends TestDistributedLogBase {
+    private static final Logger logger = LoggerFactory.getLogger(TestRollLogSegments.class);
+
+    @Test(timeout = 600000)
```
Comment (eolivelli) [inline stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java]: Nit: we banned test timeouts IIRC

File: stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+            try (BKDistributedLogManager dlm = (BKDistributedLogManager) createNewDLM(conf, name);
+                 BKAsyncLogWriter writer = dlm.startAsyncLogSegmentNonPartitioned()) {
+                writer.write(DLMTestUtil.getLogRecordInstance(1, 100000)).join();
+                writer.write(DLMTestUtil.getLogRecordInstance(2, 100000)).join();
+
+                extraBookies.forEach(b -> b.shutdown());
+
+                writer.write(DLMTestUtil.getLogRecordInstance(3, 100000));
+                Thread.sleep(1000);
```
Comment (eolivelli) [inline stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java]: Is there any safer (not flaky) way of waiting?

File: stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestDistributedLogBase.java (hunk 1 above)
Comment (ivankelly) [reply stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestDistributedLogBase.java]: oops, will remove

File: stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java (hunk 2 above)
Comment (ivankelly) [reply stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java]: they're all over the place in dlog. Will remove though.

File: stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java (hunk 3 above)
Comment (ivankelly) [reply stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java]: I can try. This sleep exists, to give time for for the write to error and the error handling to try to create a new ledger and then fail and mark the current log segment as null. Looks like it's exposed.

**Comment by ivankelly** [comment]
@eolivelli addressed comments. ptal

**Comment by ivankelly** [comment]
run integration tests

**Comment by ivankelly** [comment]
rerun integration tests

**Comment by ivankelly** [comment]
run bookkeeper-server bookie tests

**Comment by sijie** [comment]
@ivankelly I have a separate pull request (#2102) for fixing the issue introduced by #2066

**Comment by ivankelly** [comment]
run bookkeeper-server replication tests

**Comment by ivankelly** [comment]
rerun bookkeeper-server bookie tests

**Comment by ivankelly** [comment]
rerun integration tests

**Comment by sijie** [comment]
run integration tests

**Comment by ivankelly** [comment]
run integration tests

**Pull request merged**